packages/json-rpc-engine/src/v2/JsonRpcEngineV2.ts

```diff
@@ -0,0 +1,331 @@
+import {
+  type Json,
+  type JsonRpcRequest,
+  type JsonRpcNotification,
+  type NonEmptyArray,
+  hasProperty,
+} from '@metamask/utils';
+import deepFreeze from 'deep-freeze-strict';
+
+import { MiddlewareContext } from './MiddlewareContext';
+import {
+  isNotification,
+  isRequest,
+  JsonRpcEngineError,
+  stringify,
+} from './utils';
+import type { JsonRpcCall } from './utils';
+
+export type Next<Request extends JsonRpcCall, Result extends Json | void> = (
+  request?: Readonly<Request>,
+) => Promise<Readonly<Result> | void>;
+
+export type MiddlewareParams<
+  Request extends JsonRpcCall,
+  Result extends Json | void,
+> = {
+  request: Readonly<Request>;
+  context: MiddlewareContext;
+  next: Next<Request, Result>;
+};
+
+export type JsonRpcMiddleware<
+  Request extends JsonRpcCall = JsonRpcCall,
+  Result extends Json | void = Json | void,
+> = (
+  params: MiddlewareParams<Request, Result | void>,
+) => Readonly<Result> | void | Promise<Readonly<Result> | void>;
+
+type RequestState<Request extends JsonRpcCall, Result extends Json | void> = {
+  request: Request;
+  result: Result | void;
+};
+
+type Options<Request extends JsonRpcCall, Result extends Json> = {
+  middleware: NonEmptyArray<JsonRpcMiddleware<Request, Result | void>>;
+};
+
+type HandleOptions = {
+  context?: MiddlewareContext;
+};
+
+/**
+ * A JSON-RPC request and response processor.
+ *
+ * Give it a stack of middleware, pass it requests, and get back responses.
+ *
+ * @template Request - The type of request to handle.
+ * @template Result - The type of result to return.
+ *
+ * @example
+ * ```ts
+ * const engine = new JsonRpcEngineV2({
+ *   middleware,
+ * });
+ *
+ * try {
+ *   const result = await engine.handle(request);
+ *   // Handle result
+ * } catch (error) {
+ *   // Handle error
+ * }
+ * ```
+ */
+export class JsonRpcEngineV2<Request extends JsonRpcCall, Result extends Json> {
```

Comment:
Nit: Thoughts on making `JsonRpcCall` and `Json` the defaults here as well? That would save us from having to explicitly set the broadest type parameters in `JsonRpcServer`.

So long as it still infers these types correctly from the constructor params, setting a default seems more ergonomic.

Reply:
7897c2cc7